File: src/java/main/src/main/java/com/tencent/wcdb/chaincall/Select.java
// Created by chenqiuwen on 2023/5/2.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.wcdb.chaincall;

import com.tencent.wcdb.base.WCDBException;
import com.tencent.wcdb.core.Handle;
import com.tencent.wcdb.core.PreparedStatement;
import com.tencent.wcdb.orm.Field;
import com.tencent.wcdb.winq.Expression;
import com.tencent.wcdb.winq.ExpressionConvertible;
import com.tencent.wcdb.winq.OrderingTerm;
import com.tencent.wcdb.winq.StatementSelect;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class Select<T> extends ChainCall<StatementSelect> {
    private Field<T>[] fields = null;

    public Select(@NotNull Handle handle, boolean needChanges, boolean autoInvalidateHandle) {
        super(handle, needChanges, autoInvalidateHandle);
        statement = new StatementSelect();
    }

    /**
     * WINQ interface for SQL.
     *
     * @param fields The column results to be selected.
     * @return this
     */
    @SafeVarargs
    @NotNull
    public final Select<T> select(@NotNull Field<T>... fields) {
        this.fields = fields;
        statement.select(fields);
        return this;
    }

    /**
     * WINQ interface for SQL.
     *
     * @param condition condition.
     * @return this.
     */
    @NotNull
    public Select<T> where(@Nullable Expression condition) {
        statement.where(condition);
        return this;
    }

    /**
     * WINQ interface for SQL.
     *
     * @param order order term.
     * @return this.
     */
    @NotNull
    public Select<T> orderBy(@Nullable OrderingTerm order) {
        statement.orderBy(order);
        return this;
    }

    /**
     * WINQ interface for SQL.
     *
     * @param orders order list.
     * @return this.
     */
    @NotNull
    public Select<T> orderBy(@Nullable OrderingTerm... orders) {
        statement.orderBy(orders);
        return this;
    }

    /**
     * WINQ interface for SQL.
     *
     * @param count limit count.
     * @return this.
     */
    @NotNull
    public Select<T> limit(long count) {
        statement.limit(count);
        return this;
    }

    /**
     * WINQ interface for SQL.
     *
     * @param count limit expression.
     * @return this.
     */
    @NotNull
    public Select<T> limit(@Nullable ExpressionConvertible count) {
        statement.limit(count);
        return this;
    }

    /**
     * WINQ interface for SQL.
     *
     * @param offset offset number.
     * @return this.
     */
    @NotNull
    public Select<T> offset(long offset) {
        statement.offset(offset);
        return this;
    }

    /**
     * WINQ interface for SQL.
     *
     * @param offset offset expression.
     * @return this.
     */
    @NotNull
    public Select<T> offset(@Nullable ExpressionConvertible offset) {
        statement.offset(offset);
        return this;
    }

    /**
     * WINQ interface for SQL.
     *
     * @param tableName The name of the table to query data from.
     * @return this.
     */
    @NotNull
    public Select<T> from(@NotNull String tableName) {
        statement.from(tableName);
        return this;
    }

    /**
     * Get first selected object.
     *
     * @return a selected object.
     * @throws WCDBException if any error occurs.
     */
    @Nullable
    public T firstObject() throws WCDBException {
        return firstObject(Field.getBindClass(fields));
    }

    /**
     * Get first selected object.
     *
     * @param cls A derived class of java orm class.
     * @return The selected derived class object.
     * @throws WCDBException if any error occurs.
     */
    @Nullable
    public <R extends T> R firstObject(Class<R> cls) throws WCDBException {
        R ret = null;
        PreparedStatement preparedStatement = null;
        try {
            preparedStatement = prepareStatement();
            preparedStatement.step();
            if (!preparedStatement.isDone()) {
                ret = preparedStatement.getOneObject(fields, cls);
            }
        } finally {
            if (preparedStatement != null) {
                preparedStatement.finalizeStatement();
            }
            invalidateHandle();
        }
        return ret;
    }

    /**
     * Get all selected objects.
     *
     * @return A list of objects.
     * @throws WCDBException if any error occurs.
     */
    @NotNull
    public List<T> allObjects() throws WCDBException {
        return allObjects(Field.getBindClass(fields));
    }

    /**
     * Get all selected objects.
     *
     * @param cls A derived class of java orm class.
     * @return All selected derived class object.
     * @throws WCDBException if any error occurs.
     */
    @NotNull
    public <R extends T> List<R> allObjects(Class<R> cls) throws WCDBException {
        List<R> ret;
        PreparedStatement preparedStatement = null;
        try {
            preparedStatement = prepareStatement();
            ret = preparedStatement.getAllObjects(fields, cls);
        } finally {
            if (preparedStatement != null) {
                preparedStatement.finalizeStatement();
            }
            invalidateHandle();
        }
        return ret;
    }

    private PreparedStatement prepareStatement() throws WCDBException {
        return handle.preparedWithMainStatement(statement);
    }
}

File: src/rust/wcdb/src/chaincall/select.rs
use crate::base::wcdb_exception::WCDBResult;
use crate::chaincall::chain_call::{ChainCall, ChainCallTrait};
use crate::core::handle::Handle;
use crate::core::prepared_statement::PreparedStatement;
use crate::orm::field::Field;
use crate::winq::expression::Expression;
use crate::winq::ordering_term::OrderingTerm;
use crate::winq::statement::StatementTrait;
use crate::winq::statement_select::{StatementSelect, StatementSelectSelectParam};
use std::sync::Arc;

pub struct Select<'a, T> {
    chain_call: ChainCall<'a, StatementSelect>,
    fields: Vec<&'a Field<T>>,
}

impl<'a, T> ChainCallTrait for Select<'a, T> {
    fn update_changes(&self) -> WCDBResult<()> {
        self.chain_call.update_changes()
    }

    fn get_statement(&self) -> &dyn StatementTrait {
        self.chain_call.get_statement()
    }
}

impl<'a, T> Select<'a, T> {
    pub fn new(
        handle: Handle<'a>,
        need_changes: bool,
        auto_invalidate_handle: bool,
    ) -> Select<'a, T> {
        Select {
            chain_call: ChainCall::new(
                StatementSelect::new(),
                handle,
                need_changes,
                auto_invalidate_handle,
            ),
            fields: Vec::new(),
        }
    }

    // pub fn select(mut self, fields: Vec<&'a Field<T>>) -> Self {
    //     self.fields = fields;
    //     self.chain_call
    //         .statement
    //         .select(&self.fields);
    //     self
    // }

    pub fn select(mut self, fields: Vec<&'a Field<T>>) -> Self {
        self.fields = fields;
        self.chain_call.statement.select(&self.fields);
        self
    }


    pub fn r#where(self, condition: &Expression) -> Self {
        self.chain_call.statement.r#where(condition);
        self
    }

    pub fn order_by(self, order: OrderingTerm) -> Self {
        self.chain_call.statement.order_by(&vec![order]);
        self
    }

    pub fn order_by_vec(self, order_vec: &Vec<OrderingTerm>) -> Self {
        self.chain_call.statement.order_by(order_vec);
        self
    }

    pub fn limit(self, count: i64) -> Self {
        self.chain_call.statement.limit(count);
        self
    }

    pub fn offset(self, count: i64) -> Self {
        self.chain_call.statement.offset(count);
        self
    }

    pub fn from(self, table_name: &str) -> Self {
        // self.chain_call.statement.from(&vec![table_name.to_string()]);
        self
    }

    pub fn first_object(&self) -> WCDBResult<Option<T>> {
        self.first_object_by_class()
    }

    pub fn first_object_by_class(&self) -> WCDBResult<Option<T>> {
        let prepared_statement = self.prepare_statement()?;
        prepared_statement.step()?;
        let mut ret = Ok(None);
        if !prepared_statement.is_done() {
            ret = prepared_statement.get_one_object(&self.fields);
        }
        prepared_statement.finalize_statement();
        self.chain_call.invalidate_handle();
        ret
    }

    pub fn all_objects(&self) -> WCDBResult<Vec<T>> {
        self.all_objects_by_class()
    }

    pub fn all_objects_by_class(&self) -> WCDBResult<Vec<T>> {
        let prepared_statement = self.prepare_statement()?;
        let ret = prepared_statement.get_all_objects(&self.fields);
        prepared_statement.finalize_statement();
        self.chain_call.invalidate_handle();
        ret
    }

    fn prepare_statement(&self) -> WCDBResult<Arc<PreparedStatement>> {
        self.chain_call
            .handle
            .prepared_with_main_statement(&self.chain_call.statement)
    }
}
